File: backend/internal/models/graph.go
package models

import (
	"backend/internal"
	"bytes"
	"encoding/json"
	"fmt"
	"time"
)

type NodeType string

const (
	StartNode        NodeType = "start"
	ManualNode       NodeType = "manual"
	CodeGateNode     NodeType = "code"
	LocationGateNode NodeType = "location"
	RewardNode       NodeType = "reward"
)

var nodeDataDecoders = map[NodeType]func(json.RawMessage) (any, error){
	LocationGateNode: internal.DecodeInto[LocationData],
	CodeGateNode:     internal.DecodeInto[CodeData],
	ManualNode:       internal.DecodeInto[ManualData],
	RewardNode:       internal.DecodeInto[RewardData],
}

type NodePosition struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type Node struct {
	ID      string   `json:"id"`
	GraphID string   `json:"-"`
	Type    NodeType `json:"type"`

	Title       string  `json:"title"`
	Description *string `json:"description,omitempty"`

	UIPosition *NodePosition `json:"ui_position,omitempty"`

	Data any `json:"data,omitempty"`

	CreatedAt  time.Time  `json:"created_at"`
	UpdatedAt  time.Time  `json:"updated_at"`
	UnlockedAt *time.Time `json:"unlocked_at,omitempty"`
}

// UnmarshalJSON implements custom unmarshaling for Node to handle dynamic Data field.
// It uses the nodeDataDecoders map to decode the Data field based on the Node Type.
func (n *Node) UnmarshalJSON(raw []byte) error {
	type nodeAlias Node

	aux := struct {
		*nodeAlias
		Data json.RawMessage `json:"data,omitempty"`
	}{
		nodeAlias: (*nodeAlias)(n),
	}

	if err := json.Unmarshal(raw, &aux); err != nil {
		return err
	}

	// If no data (or null/empty), keep nil
	if len(bytes.TrimSpace(aux.Data)) == 0 || bytes.Equal(bytes.TrimSpace(aux.Data), []byte("null")) {
		n.Data = nil
		return nil
	}

	decoder, ok := nodeDataDecoders[n.Type]
	if !ok {
		return nil
	}

	decodedData, err := decoder(aux.Data)
	if err != nil {
		return fmt.Errorf("failed to decode node data for type %s: %w", n.Type, err)
	}

	n.Data = decodedData
	return nil
}

type NewGraphRequest struct {
	Title       string    `json:"title"`
	Description string    `json:"description"`
	StartingAt  time.Time `json:"starting_at"`
}

type Edge struct {
	ID      string `json:"id"`
	GraphID string `json:"-"`
	From    string `json:"from"`
	To      string `json:"to"`

	ChoiceLabel string `json:"choice_label,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type Viewport struct {
	X    float64 `json:"x"`
	Y    float64 `json:"y"`
	Zoom float64 `json:"zoom"`
}

type GraphStatus string

const (
	GraphNotStarted GraphStatus = "not_started"
	GraphInProgress GraphStatus = "in_progress"
	GraphCompleted  GraphStatus = "completed"
)

type Graph struct {
	ID string `json:"id"`

	Title       string `json:"title"`
	Description string `json:"description,omitempty"`

	Status GraphStatus `json:"status,omitempty"`

	StartingAt time.Time `json:"starting_at"`

	Viewport *Viewport `json:"viewport,omitempty"`

	Nodes *[]Node `json:"nodes,omitempty"`
	Edges *[]Edge `json:"edges,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type UnlockResult struct {
	UnlockedNode Node   `json:"unlocked_node"`
	NewNodes     []Node `json:"new_nodes"`
	NewEdges     []Edge `json:"new_edges"`
}

File: client/src/lib/types/graph.ts
export type GraphStatus = 'not_started' | 'in_progress' | 'completed';

export interface GraphSummary {
	id: string;
	title: string;
	description?: string;
	status?: GraphStatus;
	starting_at: Date;
	created_at: Date;
	updated_at: Date;
}

/** Full graph with accessible nodes and edges - returned by GET /api/graphs/:id */
export interface GraphDetail {
	id: string;
	title: string;
	description?: string;
	starting_at: string; // ISO 8601 date string
	nodes: AnyNode[];
	edges: Edge[];
	created_at: string; // ISO 8601 date string
	updated_at: string; // ISO 8601 date string
}

export enum NodeType {
	START = 'start',
	LOCATION = 'location',
	CODE = 'code',
	MANUAL = 'manual',
	TIME = 'time',
	REWARD = 'reward',
}

export enum RewardType {
	IMAGE = 'image',
	VIDEO = 'video',
	CALENDAR = 'calendar',
	WALLET = 'wallet',
	FAVOUR = 'favour',
	FILE = 'file',
	MARKDOWN = 'markdown',
	URL = 'url',
}

export interface Node<Type extends NodeType = NodeType, Data = undefined> {
	id: string;
	type: Type;

	title: string;
	description?: string | null;

	data?: Data;

	unlocked_at?: string | null; // ISO 8601 date string

	created_at: string; // ISO 8601 date string
	updated_at: string; // ISO 8601 date string
}

export type StartNode = Node<NodeType.START>;
export interface LocationHint {
	latitude: number;
	longitude: number;
	radius_m: number;
}

export type LocationNode = Node<NodeType.LOCATION, {
	radius_m: number;
	hint?: LocationHint | null;
}>;
export type CodeNode = Node<NodeType.CODE, {
	code: string;
}>;
export type ManualNode = Node<NodeType.MANUAL, {
	instructions: string;
	unlocked_at?: string | null; // ISO 8601 date string
}>;
export type TimeNode = Node<NodeType.TIME, {
	unlock_at: string; // ISO 8601 date string (UTC)
}>;
export interface FileInfo {
	id: string;
	filename: string;
	mime_type: string;
	size_bytes: number;
	url: string;
	url_expires_at?: string;
}

export type RewardNode = Node<NodeType.REWARD, {
	reward_type: RewardType;
	payload: string;
	give_favours: number;
	file?: FileInfo;
}>;

export type AnyNode = StartNode | LocationNode | CodeNode | ManualNode | TimeNode | RewardNode;

export interface Edge {
	id: string;
	from: string;
	to: string;
	choice_label?: string;
	created_at: string; // ISO 8601 date string
	updated_at: string; // ISO 8601 date string
}
